JDK-8263411: Convert jshell tool to use Stream.toList()

igraves opened the pull request:
This converts jshell from using `Stream.collect(Collectiors.toList())` to `Stream.toList()` - an immutable list with better performance characteristics. Local inspection only turned up one place that was mutating a resulting list and that has been refactored. 

This work is a subtask to: https://bugs.openjdk.java.net/browse/JDK-8260559

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263411](https://bugs.openjdk.java.net/browse/JDK-8263411): Convert jshell tool to use Stream.toList()


### Reviewers
 * [Jan Lahoda](https://openjdk.java.net/census#jlahoda) (@lahodaj - **Reviewer**) ⚠️ Review applies to bf0f3cfb6940674b3aa2cd88d5ddd533cfb90dd1


### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/2979/head:pull/2979`
`$ git checkout pull/2979`

To update a local copy of the PR:
`$ git checkout pull/2979`
`$ git pull https://git.openjdk.java.net/jdk pull/2979/head`

stuart-marks commented on src/jdk.jshell/share/classes/jdk/jshell/SnippetMaps.java:
```diff
@@ -164,18 +164,17 @@ String fullClassNameAndPackageToClass(String full, String pkg) {
-        for (String ipkg : pkgs) {
+                               .map(isi -> isi.fullname.substring(0, isi.fullname.lastIndexOf(".")));
+        pkgs = Stream.concat(Stream.of("java.lang"), pkgs);
+        for (String ipkg : pkgs.toList()) {
```
The original code is strictly speaking in error, as it's modifying the list returned from `Collectors.toList()` which is not guaranteed to be mutable. (Obviously, this isn't enforced.) Instead of concatenating the element to the head of the stream and using `Stream.toList()`, though, it might be just as easy to convert the original code to use `collect(Collectors.toCollection(ArrayList::new))` and leave the insertion of "java.lang" at the front at before. Might make things clearer as to what's going on.

I'll defer to the jshell team on their preference here.

lahodaj reviewed (approved):
Overall looks good to me. Please see two comments inline.

forax commented on src/jdk.jshell/share/classes/jdk/internal/jshell/tool/ConsoleIOContext.java:
```diff
@@ -597,13 +597,15 @@ public Result perform(String text, int cursor) throws IOException {
                                .map(Suggestion::continuation)
                                .distinct()
-                               .collect(Collectors.toList());
+                               .map(s -> (CharSequence) s)
```
>  that should make things clearer, I think. (And the `List` will be effectively unmodifiable by type, not only at runtime.)
Jan,
don't forget that you can always do a `add(null)` or `clear()` on a List<? extends CharSequence>, so there is no way in Java to specify an unmodifiable type for a java.util.List.

igraves replied:
Updated to match the suggestions RE the `toShow` assignments.

igraves commented:
Updated mostly in line with @lahodaj 's suggestions. I removed one explicit map-cast and replaced it with an explicit type argument to guide the type checker to the right widening.

igraves commented:
@lahodaj if you're good with these changes I can go ahead and integrate so you can sponsor.

lahodaj commented on src/jdk.jshell/share/classes/jdk/internal/jshell/tool/JShellTool.java:
```diff
@@ -1687,7 +1686,8 @@ public boolean matchesType() {
                                 }
                             })
-                            .collect(toList());
+                            .map(e -> (Suggestion) e)
```
I guess I'd get rid of the cast here as well. Otherwise looks good!

igraves replied:
Ah yes, missed that one. Done!

lahodaj commented:
Looks good!

lahodaj commented:
/sponsor

amaembo commented:
Sorry to be late but the change in SnippetMaps.java looks really confusing to me:

```java
        return Stream.concat(Stream.of("java.lang"), pkgs)
                     .filter(ipkg -> !ipkg.isEmpty() && ipkg.equals(pkg))
                     .map(ipkg -> full.substring(pkg.length() + 1))
                     .findFirst()
                     .orElse(full);
```
The elements in the stream after `map` do not depend on the previous stream content. To me, it looks a misuse of stream API. Here, `anyMatch` scenario would look better:
```java
        return Stream.concat(Stream.of("java.lang"), pkgs)
                     .anyMatch(ipkg -> !ipkg.isEmpty() && ipkg.equals(pkg)) 
                     ? full.substring(pkg.length() + 1) : full;
```

Also, it looks like the emptiness check could be applied to `pkg` variable, rather than to every stream element (as we require equality of `ipkg` and `pkg` after that:

```java
        return !pkg.isEmpty() && Stream.concat(Stream.of("java.lang"), pkgs).anyMatch(pkg::equals) 
                   ? full.substring(pkg.length() + 1) : full;
```

Sorry if I'm missing something.

igraves commented:
@amaembo Since the issue is now closed and these suggestions don't have to do with correctness, I'd suggest creating a new issue and PR to put these proposed improvements forward.

amaembo commented:
@igraves thanks, filed: https://bugs.openjdk.java.net/browse/JDK-8264221. I'm not sure about type and tags.

amaembo commented:
@igraves could you please take a look? https://github.com/openjdk/jdk/pull/3209 Thanks!
